docs/dev/development/attestation-internals.rst

```diff
@@ -0,0 +1,340 @@
+which can then be matched against the project's registered Trusted Publishers
+during the verification process.
+
+Envelope
+^^^^^^^^
+
+The ``envelope`` key contains two components:
+
+* The ``statement``, which contains the core, signed-over in-toto Statement:
+
+  .. code-block:: bash
+
+    jq -r .envelope.statement < /tmp/attestation.json | base64 -d
+
+  yielding:
+
+  .. code-block:: json
+
+    {
+      "_type": "https://in-toto.io/Statement/v1",
+      "subject": [
+        {
+          "name": "pypi_attestation_models-0.0.4a2.tar.gz",
+          "digest": {
+            "sha256": "c9709ce6fd5b67b59b4a28758cf14d3f411803c4b89b6068b1f1a8e4ee94c8ef"
+          }
+        }
+      ],
+      "predicateType": "https://docs.pypi.org/attestations/publish/v1",
+      "predicate": {}
+    }
+
+* The ``signature``, which contains the base64-encoded signature over
+  ``statement``.
+
+  ``signature`` can be verified using the public key bound within
+  ``verification_material.certificate``, fully linking the attestation back to
+  the identity that produced it.
+
+  The signing process itself is not "bare": instead of directly signing over
+  ``statement``, the payload is computed using the `DSSE PAE encoding`_:
+
+  .. code-block::
+
+    SIGNATURE = Sign(PAE(UTF8(PAYLOAD_TYPE), SERIALIZED_BODY))
+
+  where:
+
+  * ``PAYLOAD_TYPE`` is fixed as ``application/vnd.in-toto+json``
+  * ``SERIALIZED_BODY`` is the JSON-encoded ``statement``, per above
+  * ``PAE`` is the "pre-authentication encoding", defined as:
+
+    .. code-block::
+
+      PAE(type, body) = "DSSEv1" + SP + LEN(type) + SP + type + SP + LEN(body) + SP + body
+      +               = concatenation
+      SP              = ASCII space [0x20]
+      "DSSEv1"        = ASCII [0x44, 0x53, 0x53, 0x45, 0x76, 0x31]
+      LEN(s)          = ASCII decimal encoding of the byte length of s, with no leading zeros
+
+  Thus, the actual signed-over payload roughly resembles:
+
+  .. code-block::
+
+    DSSEv1 28 application/vnd.in-toto+json 272 {"_type":"https://in-toto.io/Statement/v1","subject":[{"name":"pypi_attestation_models-0.0.4a2.tar.gz","digest":{"sha256":"c9709ce6fd5b67b59b4a28758cf14d3f411803c4b89b6068b1f1a8e4ee94c8ef"}}],"predicateType":"https://docs.pypi.org/attestations/publish/v1","predicate":{}}
+
+"Why is the ``predicate`` empty?"
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You may have noticed that the in-toto Statement above contains a
+predicate of type ``https://docs.pypi.org/attestations/publish/v1``, but with an
+empty ``predicate`` body (``{}``).
+
+This is intentional! A publish attestation **does not require** a custom
+predicate, since all of the state associated with a Trusted Publisher
+is fully encapsulated in the ``verification_material.certificate`` being
+used to verify the ``envelope.statement``'s signature.
+
+Verifying an attestation object
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Attestation object verification is described at a high level in :pep:`740`.
+
```

Comment:
```suggestion

Users are **strongly discouraged** from implementing the steps below in an
ad-hoc manner, since they involve error-prone X.509 and transparency log
operations. Instead, we **strongly encourage** integrators to use
either `pypi-attestation-models`_ or `sigstore-python`_'s pre-existing APIs
for attestation manipulation, signing, and verification.

```